### PR TITLE
[TOPIC-GPIO] validate pin ordinals in wrapper functions

### PR DESCRIPTION
**Pull request by pabigot**
Extend the driver data structure with a field that specifies the number of pins present on the device.  Document the fields and who is responsible for maintaining them.
    
Update all configuration functions for specific pins to return an error if the pin is not supported.
    
Update all set/get functions for specific pins to assert if the pin is not supported.

Fix const qualifier on driver data pointers.

Driver Update Status:

Driver | Updated |Run-tested
-- | -- | -- 
cc13xx_cc26xx | :heavy_check_mark: | 
cc2650 | :heavy_check_mark: | 
cc32xx | :heavy_check_mark: | 
cmsdk_ahb | :heavy_check_mark: | 
dw | :heavy_check_mark: | 
esp32 | :heavy_check_mark: | esp32
gecko | :white_check_mark: | efr32mg_sltb004a
ht16k33 | :heavy_check_mark: | 
imx | :white_check_mark: | 
intel_apl | :white_check_mark: | 
mchp_xec | :heavy_check_mark: | 
mcux | :heavy_check_mark: | frdm_k64f
mcux_igpio | :heavy_check_mark: | mimxrt1060_evk
mcux_lpc | :heavy_check_mark: | lpcxpresso55x69_cpu0 *requires rev A2 board*
mmio32 | :heavy_check_mark: | 
nrfx | :heavy_check_mark: | nrf52840_pca10056
pca95xx | :heavy_check_mark: | 
rv32m1 | :heavy_check_mark: | rv32m1_vega_ri5cy
sam | :heavy_check_mark: | sam_e70_xplained
sam0 | :heavy_check_mark: | atsamd21_xpro
sifive | :heavy_check_mark: | hifive1_revb *test hangs, timer issues?*
stellaris | :heavy_check_mark: | 
stm32 | :heavy_check_mark: | nucleo_l476rg
sx1509b | :heavy_check_mark: | nrf52_pca20020

* :heavy_check_mark: probably good
* :white_check_mark: see commit message for limitations

Background: This implements a discussion in slack `#gpio` 2019-10-09T16:04Z quoted below:

> OK, so before doing pin validation in the wrapper functions we need to agree on what's being validated.  "Simple Poll" doesn't seem to work in this channel, so up-vote the options you want.
>* A pin index that is not less than the number of pins should cause -EINVAL to be returned *received 3+2=5 thumbs up*
>* A (port) mask that has bits set that correspond to non-existing pins should cause -EINVAL to be returned *received no votes*
>* A (port) value that has bits set that correspond to non-existing pins should cause -EINVAL to be returned *received no votes*
>
> Consider that it may be useful to pass -1 as a value or mask to indicate all pins, which is why I'm voting only for the first.

### CODE REVIEW

**Comment by galak** [comment]
This works well with the recently added DTS property `ngpios`.

**Comment by pabigot** [comment]
Explicitly marked DNM: between two commits in this every driver needs to be updated to set the ngpios field.

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: Number of pins can be determined during the compile time and is constant. We could put it inside `dev->config->config_info` structure? That saves 4 bytes of RAM. It is however more work since we currently don't expose `dev->config->config_info` at the API level.

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: A constant pointer to mutable data is what we want. The pattern is common in the driver code, typically encoded by `DEV_DATA` macro. We could possibly do
```
const struct gpio_driver_data *const data =
```
but I don't think it's worth the trouble. We don't do it inside the drivers even if the function accesses `driver_data` read only.

**Review by mnkp — CHANGES_REQUESTED** [review]
@pabigot thanks for the PR and sorry for a long delay reviewing it.

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: We'd also have to define a common structure that everybody would have to put at the start of their private config structure.   Since we already do that for data, that's not too far a stretch, but it touches every driver.

It also can't be done until the work started in #20034 is [completed](https://github.com/zephyrproject-rtos/zephyr/pull/20034#discussion_r338619559).

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: > A constant pointer to mutable data is what we want.

Well, no, it's what you want.  I think it's bizarre and unjustified, though I do see that some (not all) `DEV_DATA` macros create a const value referencing mutable data.  I'm having difficulty understanding why; the only value I can is it would stop somebody from doing:

    DEV_DATA(drv) = some_value;

I don't normally make the pointers const because people get confused and it doesn't really matter if somebody re-uses the variable with a different value, but I don't object to putting that back.  I didn't do it here because most driver adaptation headers don't do it, including other uses in this one.

I certainly do create pointers to `const` objects when I have no intention of modifying the object contents.  It lets the maintainer understand my intent.  

Are you absolutely sure you want const pointers to mutable data when the data will not be changed?  If so, I'll drop that cleanup commit and let you make the header consistent.

**Comment by mnkp** [comment]
@pabigot Let's discuss open points in the API meeting.

**Comment by carlescufi** [comment]
@pabigot as per API meeting, there is a strong preference for saving memory in RAM, meaning that the `num_pins` belongs in the `config` structure instead.

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: Moved to common config structure, some drivers updated.

File: include/drivers/gpio.h
Comment (pabigot) [reply include/drivers/gpio.h]: Used `const` qualifier on variable declaration.

**Comment by pabigot** [comment]
I've updated this with support for storing the number of pins in the static config structure, and have updated several drivers.  I've tested nrfx and sx1509b.  Note that once merged any drivers that have not been updated will fail because they'll be misinterpreting the config structure.

I'm willing to continue converting the drivers so they're all in this PR but I won't be able to test many of them, and I'll be using the `DT_INST_#_COMPAT_NGPIOS` spelling of the devicetree property.  So there may be per-driver fixups needed later.

Please provide any feedback on the general approach and I'll resume updating drivers later this week.

File: scripts/dts/edtlib.py
Comment (ulfalizer) [inline scripts/dts/edtlib.py]: Might be better to just remove this check and still let the checks below run when there's a default.

**Review by ulfalizer — CHANGES_REQUESTED** [review]

File: scripts/dts/edtlib.py
Comment (pabigot) [reply scripts/dts/edtlib.py]: Could you explain that?  If the property is required the default should be superseded by the value provided in the devicetree source, so there's no point continuing to validate the default.  In particular, it might not be compatible with the type of the property if that changed when the property was made `required` in an override.

Comment (ulfalizer) [reply scripts/dts/edtlib.py]: Nice to flag stuff like this, as it probably wasn't intended:

```
foo:
    required: true
    type: int
    default: "cool string that doesn't make sense"
```

Comment (pabigot) [reply scripts/dts/edtlib.py]: It would be, but I don't know how to do that, and I'm not going to fuss with it.

I've opened #21181 for this.  This PR will be blocked until that's resolved.  If the decision is that we have to update the SOC bindings for every target platform, then that's what we'll do.

**Comment by pabigot** [comment]
Blocked by #21181.

File: scripts/dts/edtlib.py
Comment (ulfalizer) [reply scripts/dts/edtlib.py]: Would just need to do this I think, removing all other changes:

```diff
diff --git a/scripts/dts/edtlib.py b/scripts/dts/edtlib.py
index 9877f1157e..84532fca54 100644
--- a/scripts/dts/edtlib.py
+++ b/scripts/dts/edtlib.py
@@ -1604,11 +1604,6 @@ def _check_prop_type_and_default(prop_name, prop_type, required, default,
     if default is None:
         return
 
-    if required:
-        _err("'default:' for '{}' in 'properties:' in {} is meaningless in "
-             "combination with 'required: true'"
-             .format(prop_name, binding_path))
-
     if prop_type in {"boolean", "compound", "phandle", "phandles",
                      "phandle-array"}:
         _err("'default:' can't be combined with 'type: {}' for '{}' in "
```

Comment (pabigot) [reply scripts/dts/edtlib.py]: Ah; sorry, I thought you were giving an example where the property type changed in an override, which was a case I was concerned about.  OK, I've just removed the check completely.

**Review by ulfalizer — APPROVED** [review]
Now checks `default:` against `type:` for `required: true`.

**Comment by karstenkoenig** [comment]
Sorry I just stumbled upon this due to the discussion in #gpio
Is the main purpose of this PR to check if a pin is valid? Because just comparing against the number of pins on the port is often not enough, at least in certain package variants there might be gaps. An example would be the STM32L053 (pretty random pick from historic memory), in the 64pin packages it has PC0-PC15, where as the 48pin package only has PC13-PC15. Wouldn't a bitmask of valid pins work just as well and consume just as much space? It'd be more expensive to check for but would be more flexible.

**Comment by pabigot** [comment]
I think the purpose is indeed to detect and reject invalid pins and reject them.  I wasn't aware of controllers that had gaps; for those we do need something like a bitmask.  @galak introduced `ngpios`, maybe it wasn't the right solution.

In any case, from my perspective the whole PR should be stalled until we figure out what the right solution is.  At this point every driver is doing its own validation internally (or not validating at all); maybe we should stick with that.

**Comment by karstenkoenig** [comment]
I believe a pure range based check will not be generic enough to deliver the promise of validating if a pin is existing. A bitmask might pull it off, I can't think of a case where it wouldn't, but maintaining bitmasks by package certainly ain't great and not sure if a 32bit mask would work for all platforms though.
We had a similar situation on the CAN api where drivers performed different levels of input sanitation for an invalid frame length, creating the situation that one can not rely on input validation at all. It got resolved by the test being implemented everywhere, but it's not really mandated by the interface description. A discussion about input sanitation might make sense for Zephyr in general, because input sanitation can be a bottomless hole, yet for could be argued for in a functional safety aimed RTOS.

**Comment by erwango** [comment]
I do agree with @karstenkoenig a bitmap would be the right thing to do here.
Though providing the information is not an easy thing to do (at least on STM32 side) since it would consist in adding a new level in STM32 dtsi hierarchy (maybe we can think of something smarter, but in any case this requires careful and complete study).
Besides, since information is not there today, validation cannot be done in a better way inside the driver. Last, this proposal, even if not perfect is better than current solution anyway.

So I'd propose to keep this improvement for a next evolution, once we're ready to provide full package information.

**Comment by carlescufi** [comment]
As per the API meeting the general direction of this PR is acknowledged as correct, and support for reserved ranges may be added later on top of the functionality provided here.

**Comment by pabigot** [comment]
Based on the API meeting yesterday, and the subsequent introduction of #21277, I'm going to proceed under the assumption that eventually something will be producing a bitmask of valid pins, and will change the config value and the verification to be based on that interface.  For now I'll initialize the mask to `BIT(ngpios) - 1U`, but presumably eventually the `gpio-ranges` pairs will be translated into something that produces that bitmask for the controller and exposes it as a generated macro value.

File: drivers/gpio/gpio_stm32.c
Comment (erwango) [inline drivers/gpio/gpio_stm32.c]: I actually spent some time to understand how substracting 1 to gpio_port_pins_t could work. Anyway to disambiguate ?

Then, STM32 GPIO ports actually control 16 bits, so we should use `GPIO_AVAILABLE_PINS` macro, checking ngpios value. Bright side is that code will be ready for update to a range version.
Adding the whole fixup code is a bit cumbersome, so I'm preparing a commit to spare you the effort.

Last, I've tried to set `ngpios` by default to 16 `st,stm32-gpio.yaml`, but it is actually not possible to override the 32 value defined in `gpio-controller`. Is there a strong requirement to prevent this?
I'm not quite in favor of spreading dts values in bindings, but this is indeed handy for this kind of things.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_stm32.c
Comment (pabigot) [reply drivers/gpio/gpio_stm32.c]: I suppose `(gpio_port_pins_t)(-1)` might provide a clue where necessary.

If the GPIO ports control 16 bits, then this should be `GPIO_AVAILABLE_PINS(16U)`.  I could make that change.  IIRC I went with the literal because the implementation depends on fixups to get the right devicetree value and I didn't want to add all of those (I ultimately had to for IMX because reasons, but it's not good).  I'd welcome a patch that turns it into what you want.

I don't see any reason not to allow overriding an existing default; the whole dts binding description architecture is pretty ad hoc.  This PR requires and includes #21271; if that doesn't get merged soon I'll need to push for it.  If you make a PR to support default override we can include it here too.

Comment (erwango) [reply drivers/gpio/gpio_stm32.c]: +1 for `(gpio_port_pins_t)(-1)`
`GPIO_AVAILABLE_PINS(16U)` is fine with me as well. I can push an update later on to work with a dts extracted info, but all in all, I prefer this than the bindings inferred default `ngpios` value.

Comment (pabigot) [reply drivers/gpio/gpio_stm32.c]: OK, I'll queue those changes in the next update.

Comment (erwango) [reply drivers/gpio/gpio_stm32.c]: Thanks!

Comment (pabigot) [reply drivers/gpio/gpio_stm32.c]: Done.

**Review by erwango — APPROVED** [review]
Tested ok

**Review by henrikbrixandersen — APPROVED** [review]
HT16K33 changes OK.

**Comment by pabigot** [comment]
@mnkp The issues in your previous review were addressed after the meeting where we discussed them.  Please re-check.

**Review by mnkp — COMMENTED** [review]
@pabigot I believe I misunderstood your comment during the API meeting. I thought the idea was to do a rough check based on ngpios property in the API and leave a more detailed check to be implemented by the driver. The current solution with bit mask is very convenient and flexible. However, very likely it will not be future proof. And looking into the future I believe we need to do a better job to support hardware GPIO modules like the Intel one #20400 where one GPIO controller can have up to 78 pins.

To improve support for such GPIO modules we would need to add _pin_ functions to the driver API (the user GPIO API wouldn't change). In this case having check based on mask wouldn't work. We would need a solution like in Linux where they define list of valid pin ranges.

**Comment by pabigot** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/20115#discussion_r361815226 changed to assert.

For the overall approach:  https://github.com/zephyrproject-rtos/zephyr/pull/20115#issuecomment-564560462 explains the switch to bitmask.  Since the intent is to get `gpio-reserved-ranges` in soon this is the future-proof path that allows drivers that need the feature to use it.

Given the difficulty in getting this GPIO API change in (over 8 months so far since #15611), I can't see that another rework to support more than 32 pins per gpio controller will ever happen.  Nexus maps mostly allow the limitation to be hidden from developer code, as demonstrated for the up-squared platform.

**Review by mnkp — COMMENTED** [review]
> Nexus maps mostly allow the limitation to be hidden from developer code, as demonstrated for the up-squared platform.

This is a good workaround but it's still a workaround. Adding pin level functions to GPIO drivers will be significantly less work than what we are doing now. It's not going to happen tomorrow, there are more important things to work on but at some point we should to it.

We can also change pin validation code later (I have some more comments to the current implementation in this case). Indeed, we shouldn't delay the current work.

There is one more important thing which needs to be done now however. Since pin mask of available pins is stored in ROM we need to verify that - given `gpio-reserved-ranges` property - we can populate the mask during the compile time. Otherwise the concept will not work.

File: drivers/gpio/gpio_utils.h
Comment (pabigot) [reply drivers/gpio/gpio_utils.h]: No, it has to be this way.  `BIT(32)` produces an overflow when `1UL << 32` is evaluated and unsigned long is a 32-bit value.

File: drivers/gpio/gpio_stm32.c
Comment (pabigot) [reply drivers/gpio/gpio_stm32.c]: See the commit message: it's not possible to do that here, and I believe @erwango approved this as the short-term workaround.

File: drivers/gpio/gpio_utils.h
Comment (pabigot) [reply drivers/gpio/gpio_utils.h]: I can do `GPIO_PORT_PINS_MASK_FROM_NGPIOS` here, and change `available` to `port_pins_mask`.

**Comment by pabigot** [comment]
> There is one more important thing which needs to be done now however. Since pin mask of available pins is stored in ROM we need to verify that - given `gpio-reserved-ranges` property - we can populate the mask during the compile time. Otherwise the concept will not work.

That's going to have to be the responsibility of the folks doing the devicetree scripting support.  There'll have to be custom code that generates the appropriate mask for all identified pins.

**Comment by mnkp** [comment]
> > There is one more important thing which needs to be done now however. Since pin mask of available pins is stored in ROM we need to verify that - given `gpio-reserved-ranges` property - we can populate the mask during the compile time. Otherwise the concept will not work.
> 
> That's going to have to be the responsibility of the folks doing the devicetree scripting support. There'll have to be custom code that generates the appropriate mask for all identified pins.

In this case, shouldn't we generate the pin mask already, without support for `gpio-reserved-ranges`, or otherwise we will need to touch all the gpio drivers again. Also, we would make sure that this custom code is fine with devicetree folks.

**Comment by pabigot** [comment]
> In this case, shouldn't we generate the pin mask already, without support for `gpio-reserved-ranges`, or otherwise we will need to touch all the gpio drivers again. Also, we would make sure that this custom code is fine with devicetree folks.

Only drivers for controllers that have gaps in the supported pins would have to be updated: every other driver can use the mask-from-ngpios code this PR provides.  I don't want to hold this up for another release because we're stalled waiting for more devicetree tooling to get implemented.

File: drivers/gpio/gpio_utils.h
Comment (mnkp) [reply drivers/gpio/gpio_utils.h]: I misread the purpose of the code. But in this case it would be cleaner to cast to long long or uintmax_t. This will be computed during the compilation time, shouldn't influence the code size.

**Comment by mnkp** [comment]
> Only drivers for controllers that have gaps in the supported pins would have to be updated: every other driver can use the mask-from-ngpios code this PR provides.

If adding the custom code to DTS scripts to generate the mask is not a problem, then that's indeed a good approach.

File: drivers/gpio/gpio_utils.h
Comment (pabigot) [reply drivers/gpio/gpio_utils.h]: Should all be updated.

File: include/drivers/gpio/gpio_mmio32.h
Comment (mnkp) [inline include/drivers/gpio/gpio_mmio32.h]: `.port_pin_mask`

File: include/drivers/gpio.h
Comment (mnkp) [inline include/drivers/gpio.h]: Other asserts print "Unsupported pin", we could use a standard message.

File: tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c
Comment (mnkp) [inline tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c]: All the functions use ASSERT now, the testcase can be removed (or we should write a testcase that tests the ASSERT).

Comment (pabigot) [reply tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c]: Then that's wrong: the plan was to return `-EINVAL` when an invalid pin was provided to a function, per the decision recorded in the PR description.  We should either revert some of those changes to `ASSERT` or get agreement that it's OK to use only assert.  To discuss next API meeting.

Comment (pabigot) [reply tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c]: Per API meeting: test will be removed.

Comment (pabigot) [reply tests/drivers/gpio/gpio_basic_api/src/test_gpio_port.c]: done

**Review by mnkp — APPROVED** [review]

**Pull request merged**
